addAtomTypes only updates atomic number info for types existing in .dat/frcmod // hard-code EP and LP

rafwiewiora opened the pull request:
As we discussed in https://github.com/ParmEd/ParmEd/issues/486 . 

Not sure if this is the best approach here, what do you think?

rafwiewiora commented:
Thinking about this more, I'm no longer sure this is necessary.. 

The use case here would be to have this set to False in the conversion script that will be shipped with OpenMM - if people want to convert a leaprc, we won't consider the addAtomTypes section to avoid potential Atom Type clashes. 

But then - if you're not going to manually modify that leaprc, you will run into trouble anyway most likely, because the leaprc will have ions and solvent parameters and that will cause clashes etc. So you might as well rely on deleting the addAtomTypes from the leaprc if you don't want them.

Well let's see what you think!

rafwiewiora commented:
Also this is assuming that the element recognition works well by mass - it seems to. But maybe it'd be better to not throw away the element information and when the .dat is loaded, go back and delete those atom types that were not declared in the .dat?

swails commented:
Can you check that `atomic_number` still gets set correctly on the `AtomType` objects in `atom_types`.  Because if it doesn't, I don't think we should ever ignore this completely.

swails commented:
Oh, I looked at a really old window where your last comment didn't pop up.  Since `atomic_number` seems to be set from the mass by default (thinking about it more, I think it *must* be, since I added parm.dat/frcmod parsing before leaprc parsing), then addAtomTypes really doesn't do anything.  I'd almost suggest just commenting out that whole segment and just identifying virtual sites by name (EP or LP, since Amber hard-codes for those two).

rafwiewiora commented:
Looks like it all works fine: as long as the recognition of the element by mass works ok, the atomic number flows from that: https://github.com/ParmEd/ParmEd/blob/ommparams/parmed/amber/parameters.py#L551-L552

One issue I see now is Mn vs Fe - ff99SBildn for example has Fe (residue FE) at mass 55.0, masses in https://github.com/ParmEd/ParmEd/blob/master/parmed/periodic_table.py are Mn: 54.93, Fe: 55.84 - and hence residue FE gets recognized as element Mn. Not surprising given these numbers, but if we kept addAtomTypes this would be definied correctly as Fe. What to do?

Ok, so let's combine this with fixing https://github.com/ParmEd/ParmEd/issues/580 and I'm going to (depending on this Fe problem):

* comment out the addAtomTypes section
* hard code EP and LP element recognition 

One more problem I have with the elements is these weird ion things, like IB ( 'big ion w/ waters' for vacuum (Na+, 6H2O)) - which gets recognized as Xenon because of the mass. I also see IM and IP in 99SBildn - but these are recognized as Cl and Na correctly. I guess there might be other ones like this in other FFs? How can we deal with this? I'm seeing big problems if someone wants to use say IB in OpenMM, which uses the element-based template recognition for residues etc.

drroe commented:
> One issue I see now is Mn vs Fe - ff99SBildn for example has Fe (residue FE) at mass 55.0, masses in https://github.com/ParmEd/ParmEd/blob/master/parmed/periodic_table.py are Mn: 54.93, Fe: 55.84 - and hence residue FE gets recognized as element Mn. Not surprising given these numbers, but if we kept addAtomTypes this would be definied correctly as Fe. What to do?

FYI what cpptraj does (based on how Sander QM/MM code recognizes elements) is to also use name information. So in your case the `F` in `FE` would let the code know it's probably iron and not manganese.

rafwiewiora commented:
> FYI what cpptraj does (based on how Sander QM/MM code recognizes elements) is to also use name information. So in your case the F in FE would let the code know it's probably iron and not manganese.

I like that! But that sounds like writing quite a few lines, I think it might be better for now to keep the addAtomTypes and only use that info to assign elements, where available, without adding the atom types that are not also defined in a .dat or frcmod.

hainm commented on parmed/amber/parameters.py:
```diff
@@ -211,7 +211,7 @@ def __init__(self, *filenames):
 
     @classmethod
-    def from_leaprc(cls, fname):
+    def from_leaprc(cls, fname, addAtomTypes=True):
```
Hi, Is there any specific reason to keep camelCase here?

rafwiewiora replied:
I wasn't sure what you guys use here, camelCase is just strong in me from coding for OpenMM! Also that keeps the camelCase that Amber uses for this command. But anyway, this PR will not stay in this form, looks like it.

swails replied:
Except for areas where the ParmEd and OpenMM APIs overlap (like `createSystem` and the like), ParmEd uses the `snake_naming_convention`.

rafwiewiora replied:
Ok good to know!

swails commented:
> FYI what cpptraj does (based on how Sander QM/MM code recognizes elements) is to also use name information. So in your case the F in FE would let the code know it's probably iron and not manganese.

That's how I'd have done it if I wasn't a lazy purist :).  (I also knew that's how sander's QM/MM code *still* does it, even when `ATOMIC_NUMBER` is a thing in the prmtop now.)  But I didn't actually expect anything to get atomic masses *that wrong*.

swails commented:
> One issue I see now is Mn vs Fe - ff99SBildn for example has Fe (residue FE) at mass 55.0, masses in https://github.com/ParmEd/ParmEd/blob/master/parmed/periodic_table.py are Mn: 54.93, Fe: 55.84 - and hence residue FE gets recognized as element Mn. Not surprising given these numbers, but if we kept addAtomTypes this would be definied correctly as Fe. What to do?

Ugh.  Correct the stupid imprecise mass in Amber's files.

> One more problem I have with the elements is these weird ion things, like IB ( 'big ion w/ waters' for vacuum (Na+, 6H2O)) - which gets recognized as Xenon because of the mass. I also see IM and IP in 99SBildn - but these are recognized as Cl and Na correctly. I guess there might be other ones like this in other FFs? How can we deal with this? I'm seeing big problems if someone wants to use say IB in OpenMM, which uses the element-based template recognition for residues etc.

Just get rid of those atom types.  They're never used inside AMBER.  Their presence in the parm.dat files isn't really a big deal since they've been removed from the residue templates.  So it's just rubble left over that isn't hurting anyone, so it wasn't cleaned away.  Since they're not accessible from any of the Amber force fields we're trying to port, I would say they're not part of those force fields.

hainm commented:
> even when ATOMIC_NUMBER is a thing in the prmtop now.

But if this is not available in prmtop (generated by old tleap), qm/mm still make a guess based on name.

swails commented:
> But if this is not available in prmtop (generated by old tleap), qm/mm still make a guess based on name.

Yes, but the point is that sqm never uses ATOMIC_NUMBER, even if it's there.  And any prmtop that has been created in the last 5 years has ATOMIC_NUMBER present (and ParmEd has the `addAtomicNumber` command to add it if you needed to).

drroe commented:
> and ParmEd has the addAtomicNumber command to add it if you needed to

Cpptraj also adds atomic number information, so really sqm has no excuse :-)

swails commented:
You mean except the way-too-long excuse about having no time or money for any of a number of nefarious reasons?

rafwiewiora commented:
> Just get rid of those atom types. They're never used inside AMBER. Their presence in the parm.dat files isn't really a big deal since they've been removed from the residue templates. So it's just rubble left over that isn't hurting anyone, so it wasn't cleaned away. Since they're not accessible from any of the Amber force fields we're trying to port, I would say they're not part of those force fields.

Actually, there are still in the `.lib`, check out ions94.lib which is used with ff99SBildn. It has the IB residue and a CIO that uses the IP type....

rafwiewiora commented:
tagging @jchodera here for an input on these ions too, see:

> One more problem I have with the elements is these weird ion things, like IB ( 'big ion w/ waters' for vacuum (Na+, 6H2O)) - which gets recognized as Xenon because of the mass. I also see IM and IP in 99SBildn - but these are recognized as Cl and Na correctly. I guess there might be other ones like this in other FFs? How can we deal with this? I'm seeing big problems if someone wants to use say IB in OpenMM, which uses the element-based template recognition for residues etc.

and comment above quoting @swails 's reply. 

I don't think anyone would be using these in OpenMM (are these some crazy relict of the past btw?), but it might be hard picking things like this not to include, so I'm leaning towards just converting absolutely everything there is in the Amber files. But I need to sort out the element issue.

drroe commented:
> are these some crazy relict of the past btw?

Yes, I believe IM stands for "ion minus" and IP for "ion plus" (Cl- and Na+ respectively)

rafwiewiora commented:
> Yes, I believe IM stands for "ion minus" and IP for "ion plus" (Cl- and Na+ respectively)

Aha! Ok, so there's both an IM and a Cl type - IM is for the Cl- ion, Cl is from CHCl3, that makes sense. But then there's both an IP type and a Na type (described as Na+ ion in the .dat), they have the same parameters (they are just a copy), but the .lib still uses an IP, rather than Na for the Na+ ion. 

And then we have the CIO residue, which is again a Na+ ion using the IP type (though there's another Na+ residue that is the same. Wow....

swails commented:
These ion models are way out-of-date, though...  They were removed well before ff99SB was really considered deprecated (ff10 did away with these ions94.lib-style ions, and ff10 was identical to ff99SB for proteins).

The new ion parameters are *much* better than what you get with IM and IP.  I'm not sure how you guys want to handle this, but *I* will always use (and will always suggest others use) the newer water model-specific ion parameters.

jchodera commented:
I'm 100% in agreement.  These old ion parameters will [cause the ions to crystallize well below the solubility limit](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2652252/).  

We don't have a fully defined philosophy about this, but my understanding is that @peastman would prefer (1) old colde shouldn't break, and (2) we make it easy to do the right thing by default.  I would also prefer we can (3) use old parameters when needed for compatibility.  When it gets to choosing file naming conventions, we might want the default `ffxml` "bundles" to match up with the old names when possible but use updated parameter bundles, but with outdated parameters moved to something like `deprecated/` or `old/`.  We should also update the default ion names in `Modeller` as appropriate if those have changed.

rafwiewiora commented:
Ok, cool!

swails commented:
> (1) old colde shouldn't break,

This is an important concern, but I don't think the proposed changes here will affect that.  Not unless we propose *replacing* previously-converted XML files that are already in OpenMM, and I would recommend against that path (unless there's a convincing argument to do so?).

> (2) we make it easy to do the right thing by default.

To me, this is more important than the first point (although Peter may have slightly different priorities)

> (3) use old parameters when needed for compatibility.

You mean something along the lines of `ForceField('ff14sb.xml', 'really_crappy_ion_parameters.xml')`? :smiley:

Back to this PR, though -- are we still doing something with this, or do some of the newer PRs address what we're trying to fix here?

rafwiewiora commented:
Ok, I'm working on this now. Will:

* get rid of the changes as proposed in the first version of the PR
* hard-code EP and LP element recognition
* think about what to do with the Mn - Fe problem and how this ties into whether too comment out the whole addAtomTypes section

rafwiewiora commented:
Ok, I've changed this: moved addAtomTypes loading to after parameters and limited it only to overriding atomic number information for existing types, it can't add new ones. 

Also hard-coded 'EP' and 'LP' for atomic number 0. So the regex will continue not recognizing those in addAtomTypes ("" element), hence the hard-code only for parameter reading. 

How does it look?

rafwiewiora commented:
#587 supplements this with atomic number 0 --> "" element translation for ffxml writing

rafwiewiora commented:
And fixes #580

rafwiewiora commented:
Fixed failing test - previous K+ was added by addAtomTypes, but wasn't present in the dat or frcmod - so now is gone.

swails commented:
Looks good to me.  Is this ready to merge in your opinion?

rafwiewiora commented:
Yep, good to go in!

swails commented:
OK, I merged this into #588.  After Travis passes, I'll merge that in.  Then I'll merge #587 unless there are conflicts.  Good plan?  Or would you rather I swapped the merge order?

rafwiewiora commented:
Sounds like a good plan to me!
